Source/Scene/Model/MetadataPipelineStage.js

```diff
@@ -62,60 +83,225 @@ MetadataPipelineStage.process = function (
+    // GeometryPipelineStage to ensure we have the correct GLSL type
+    const modelAttribute = getAttributeByName(primitive, property.attribute);
+    const { glslType, variableName } = getAttributeInfo(modelAttribute);
+
+    return {
+      metadataVariable: sanitizeGlslIdentifier(propertyId),
+      property,
+      type: property.classProperty.type,
+      enumType: property.classProperty.enumType,
+      glslType,
+      variableName,
+      propertyStatistics: classStatistics?.properties[propertyId],
+      shaderDestination: ShaderDestination.BOTH,
+    };
+  }
+}
+
+/**
+ * Collect info about all properties of all propertyTextures, and
+ * return as a flattened Array
+ * @param {PropertyTexture[]} propertyTextures
+ * @param {Object} statistics
+ * @returns {Array<{Object}>}
+ * @private
+ */
+function getPropertyTexturesInfo(propertyTextures, statistics) {
+  if (!defined(propertyTextures)) {
+    return [];
+  }
+  return propertyTextures.flatMap((propertyTexture) =>
+    getPropertyTextureInfo(propertyTexture, statistics)
+  );
+}
+
+/**
+ * Collect info about the properties of a PropertyTexture
+ * @param {PropertyTexture} propertyTexture
+ * @param {Object} statistics
+ * @returns {Array<{Object}>}
+ * @private
+ */
+function getPropertyTextureInfo(propertyTexture, statistics) {
+  const { sanitizeGlslIdentifier } = ModelUtility;
+
+  const classId = propertyTexture.class.id;
+  const classStatistics = statistics?.classes[classId];
+
+  return Object.entries(propertyTexture.properties)
+    .map(getPropertyTexturePropertyInfo)
+    .filter(defined);
+
+  function getPropertyTexturePropertyInfo([propertyId, property]) {
+    if (!property.isGpuCompatible()) {
+      return;
+    }
+
+    return {
+      metadataVariable: sanitizeGlslIdentifier(propertyId),
+      property,
+      type: property.classProperty.type,
+      enumType: property.classProperty.enumType,
+      glslType: property.getGlslType(),
+      propertyStatistics: classStatistics?.properties[propertyId],
+      shaderDestination: ShaderDestination.FRAGMENT,
+    };
+  }
+}
+
 /**
  * Declare <type>MetadataClass structs in the shader
  * @param {ShaderBuilder} shaderBuilder The shader builder for the primitive
  * @param {Set<String>} metadataTypes The types of metadata used in the shaders
+ * @param {Object} statistics Statistics about the metadata.
  * @private
  */
-function declareMetadataTypeStructs(shaderBuilder, metadataTypes) {
+function declareMetadataTypeStructs(shaderBuilder, propertyInfos) {
+  const classTypes = new Set();
+  const statisticsTypes = new Set();
+  const enumLengths = new Set();
+
+  for (let i = 0; i < propertyInfos.length; i++) {
+    const { type, enumType, glslType, propertyStatistics } = propertyInfos[i];
+    classTypes.add(glslType);
+    if (!defined(propertyStatistics)) {
+      continue;
+    }
+    if (type === MetadataType.ENUM) {
+      enumLengths.add(enumType.values.length);
```

Comment:
:thinking: are enum values required to be a contiguous range of values in the spec? If sparse values are allowed, then this won't work in all cases. If that's the case, we should discuss this further.

Reply:
We could allow sparse values by sizing the array based on the maximum index, rather than the number of indices.

However, the [spec](https://github.com/CesiumGS/3d-tiles/blob/draft-1.1/specification/schema/Schema/enum.schema.json) also allows signed integers or strings as values.

If we really want to handle everything the spec allows, I think we would need to make the occurrences statistic a struct rather than an array, with statistics keyed on the enum name rather than the value. For the test dataset, accessing the value would look like this:

```glsl
int buildingsCount = vsInput.metadataStatistics.classification.occurrences.Buildings;
```

This would also enable statistics for fixed-length arrays of enum values. Some additional work would be required:
- Add sanitization for enum names
- Document the sanitization code, so Custom Shader authors know the resulting struct field names
- Confirm if the code actually supports array-valued metadata

We need to decide how much of the spec we want to support in one PR. The spec allows varying-length arrays, which could be very challenging to implement in WebGL1.

Reply:
```glsl
int buildingsCount = vsInput.metadataStatistics.classification.occurrences.Buildings;
```

Technicals aside -- I actually prefer this API since it resembles how you would access the value from JSON.

I also think it's important to support non-contiguous ranges of values and signed integers, which this does.

If there's one spec feature we cut from this PR, I think array-valued enums would be ok. But just to clarify one thing, varying length array properties can't have statistics. Though after re-reading the section in the spec it's not super clear.

Reply:
As discussed offline: I will drop support for `occurrences` in this PR. 

Support for `occurrences` will be added in a separate PR. We plan to expose it in the shader via a struct keyed on the enum name.